src/accelerate/utils/bnb.py

```diff
@@ -0,0 +1,390 @@
+    bnb_quantization_config.skip_modules.extend(modules_on_cpu)
+    modules_to_not_convert = bnb_quantization_config.skip_modules
+    # We add the modules we want to keep in full precision
+
+    if bnb_quantization_config.keep_in_fp32_modules is None:
+        bnb_quantization_config.keep_in_fp32_modules = []
+    keep_in_fp32_modules = bnb_quantization_config.keep_in_fp32_modules
+    modules_to_not_convert.extend(keep_in_fp32_modules)
+
+    if model.device.type != "meta":
+        # quantization of an already loaded model
+        logger.warning("It is not recommended to quantize a loaded model. There might be some memory issues.")
+        model = replace_with_bnb_layers(model, bnb_quantization_config, modules_to_not_convert=modules_to_not_convert)
+        # convert param to the right dtype
+        dtype = bnb_quantization_config.torch_dtype
+        for name, param in model.state_dict().items():
+            if any(module_to_keep_in_fp32 in name for module_to_keep_in_fp32 in keep_in_fp32_modules):
+                param.to(torch.float32)
+                if param.dtype != torch.float32:
+                    name = name.replace(".weight", "").replace(".bias", "")
+                    param = getattr(model, name, None)
+                    if param is not None:
+                        param.to(torch.float32)
+            elif torch.is_floating_point(param):
+                param.to(dtype)
+        if model.device.type == "cuda":
+            # move everything to cpu in the first place because we can't do quantization if the weights are already on cuda
+            model.to("cpu")
+            torch.cuda.empty_cache()
+        if torch.cuda.is_available():
+            model.to(torch.cuda.current_device())
+        else:
+            raise RuntimeError("No GPU found. A GPU is needed for quantization.")
+        logger.info(
+            f"The model device type is {model.device.type}. However, cuda is needed for quantization."
+            "We move the model to cuda."
+        )
+        return model
+    else:
+        if weights_location is None:
+            raise RuntimeError(
+                f"`weights_location` needs to be the folder path containing the weights of the model, but we found {weights_location} "
+            )
+        else:
+            with init_empty_weights():
+                model = replace_with_bnb_layers(
+                    model, bnb_quantization_config, modules_to_not_convert=modules_to_not_convert
+                )
+            # Make sure tied weights are tied before creating the device map.
+            model.tie_weights()
+            device_map = get_quantized_model_device_map(
+                model,
+                bnb_quantization_config,
+                device_map,
+                max_memory=max_memory,
+                no_split_module_classes=no_split_module_classes,
+            )
+            print
```

Comment:
To remove.

Reply:
removed here [0dcc440](https://github.com/huggingface/accelerate/pull/1626/commits/0dcc440d902228a57e7148198758eba6b1b09f45)